### ISSUE TITLE
Underspecified behavior of global unicode regular expressions with lastIndex inside a surrogate pair

### ISSUE THREAD

**Issue by gibson042** [tc39/ecma262#2511]
**Description:**
[RegExpBuiltinExec](https://tc39.es/ecma262/#sec-regexpbuiltinexec) invokes its matcher closure with the input string and the value derived from `lastIndex`, and [_Pattern_ semantics](https://tc39.es/ecma262/#sec-pattern) for a unicode regular expression interpret the string _str_ as a sequence of code points _Input_ and then in step e "Let _listIndex_ be the index into _Input_ of the character that was obtained from element _index_ of _str_." But it is unclear what should happen when _index_ identifies the trailing surrogate of a pair—should _listIndex_ be the position of the character derived from the pair, or should it be the position after that?

Unsurprisingly, the results of this underspecification are implementation divergence.

**eshost Output:**
```sh
$ eshost -sx 'var r = /\udf06/ug; r.lastIndex=1; var result = r.exec("\ud834\udf06\udf06");
    print(result && result.index, r.lastIndex)'
#### ChakraCore, Hermes, JavaScriptCore, Moddable XS
1 2

#### engine262
1 3

#### GraalJS, SpiderMonkey, V8
2 3
```
```sh
$ eshost -sx 'var r = /./ug; r.lastIndex=1; var result = r.exec("\ud834\udf06\udf06");
    print(result && result.index, r.lastIndex)'
#### ChakraCore, Hermes, JavaScriptCore, Moddable XS
1 2

#### engine262
1 3

#### GraalJS, SpiderMonkey, V8
0 2
```
JSC, XS, et al. interpret the trailing surrogate as if it were unpaired, but that is not consistent with closure algorithm step c ("let _Input_ be a List whose elements are the code units that are the elements of _str_"). Graal, SpiderMonkey, and V8 get that part right, but seem to vary based on the regular expression itself whether or not to check the start of the character split by `lastIndex`. And engine262 is just right out.

Closure algorithm step e should clearly be clarified, and I see two acceptable options:
1. Have _listIndex_ correspond with the position of the character split by `lastIndex`, stepping backward by one code unit.
2. Have _listIndex_ correspond with the position after that.

Option 2 makes more sense to me, but this is such an edge case that I wouldn't argue strongly against option 1.

**Comment by allenwb** [comment]
> step e "Let _listIndex_ be the index into _Input_ of the character that was obtained from element _index_ of _str_."

I originally wrote  that text. The intended meaning when  _Unicode_ is **true** was that a string _index_  value corresponding to either surrogate of a valid surrogate pair is a reference to the corresponding code point.  IOW, when _index_ identifies the trailing surrogate of a pair, _listIndex_ is the position of the character derived from the pair.

That was an intentional design discussion and generally how string indices that reference a valid trailing surrogate of a pair  are handled throughout the RegExp spec.  For example see step 12.a of [RegExpBuiltinExec](https://tc39.es/ecma262/#sec-regexpbuiltinexec)

**Comment by gibson042** [comment]
12.a unambiguously maps an index in a list of code points to the index of its first code unit in the string from which those code points were derived, so I don't see the connection. But regardless, the original author indicating that option 1 was intended is certainly strong support for it.

**Comment by anba** [comment]
This looks like a duplicate of #128.

**Comment by gibson042** [comment]
I agree, and find this consequence of option 1 mentioned in https://github.com/tc39/ecma262/issues/128#issuecomment-393748953 to be particularly interesting:
> It would be helpful to add a note that a regexp can produce a match at an index prior to lastIndex. I would think many developers have traditionally assumed that a match could only be found at, or after, lastIndex.
> 
> Code that works with a caller-provided RegExp object will suddenly break if they are given certain unicode regexps.
